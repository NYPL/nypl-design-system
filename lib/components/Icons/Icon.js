"use strict";
var __createBinding =
    (this && this.__createBinding) ||
    (Object.create
        ? function (o, m, k, k2) {
              if (k2 === undefined) k2 = k;
              Object.defineProperty(o, k2, {
                  enumerable: true,
                  get: function () {
                      return m[k];
                  },
              });
          }
        : function (o, m, k, k2) {
              if (k2 === undefined) k2 = k;
              o[k2] = m[k];
          });
var __setModuleDefault =
    (this && this.__setModuleDefault) ||
    (Object.create
        ? function (o, v) {
              Object.defineProperty(o, "default", {
                  enumerable: true,
                  value: v,
              });
          }
        : function (o, v) {
              o["default"] = v;
          });
var __importStar =
    (this && this.__importStar) ||
    function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null)
            for (var k in mod)
                if (Object.hasOwnProperty.call(mod, k))
                    __createBinding(result, mod, k);
        __setModuleDefault(result, mod);
        return result;
    };
var __importDefault =
    (this && this.__importDefault) ||
    function (mod) {
        return mod && mod.__esModule ? mod : { default: mod };
    };
Object.defineProperty(exports, "__esModule", { value: true });
const React = __importStar(require("react"));
const bem_1 = __importDefault(require("../../utils/bem"));
const arrow_xsmall_svg_1 = __importDefault(
    require("../../icons/svg/arrow-xsmall.svg")
);
const arrow_svg_1 = __importDefault(require("../../icons/svg/arrow.svg"));
const download_svg_1 = __importDefault(require("../../icons/svg/download.svg"));
const minus_svg_1 = __importDefault(require("../../icons/svg/minus.svg"));
const plus_svg_1 = __importDefault(require("../../icons/svg/plus.svg"));
const search_small_svg_1 = __importDefault(
    require("../../icons/svg/search-small.svg")
);
const search_svg_1 = __importDefault(require("../../icons/svg/search.svg"));
const allSvgs = {
    arrow_xsmall: arrow_xsmall_svg_1.default,
    arrow: arrow_svg_1.default,
    download: download_svg_1.default,
    minus: minus_svg_1.default,
    plus: plus_svg_1.default,
    search_small: search_small_svg_1.default,
    search: search_svg_1.default,
};
class Icon extends React.Component {
    constructor(props) {
        super(props);
    }
    render() {
        let {
            modifiers = [],
            blockName,
            name,
            decorative,
            role,
            title,
            desc,
            iconRotation,
        } = this.props;
        let icon_base_class = "icon";
        if (this.props.iconRotation) {
            modifiers.push(this.props.iconRotation);
        }
        let iconProps = {
            className: bem_1.default(icon_base_class, modifiers, blockName),
            role: decorative ? "img" : role,
            "aria-hidden": decorative,
            "aria-labelledby": title ? "title-" + name : undefined,
            "aria-describedby": desc ? "desc-" + name : undefined,
            title: title ? `title-${name}` : undefined,
            description: desc ? `title-${name}` : undefined,
        };
        let svg = allSvgs[name];
        return React.createElement(
            "span",
            Object.assign({}, iconProps, {
                dangerouslySetInnerHTML: { __html: svg },
            })
        );
    }
}
exports.default = Icon;
