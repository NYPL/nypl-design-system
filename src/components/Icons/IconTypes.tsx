/* eslint-disable camelcase */
export enum IconRotationTypes {
  rotate0 = "rotate-0",
  rotate90 = "rotate-90",
  rotate180 = "rotate-180",
  rotate270 = "rotate-270",
}

export enum IconNames {
  accessibility_full = "accessibility_full",
  accessibility_partial = "accessibility_partial",
  arrow = "arrow",
  check = "check",
  clock = "clock",
  close = "close",
  download = "download",
  error_filled = "error_filled",
  error_outline = "error_outline",
  headset = "headset",
  minus = "minus",
  plus = "plus",
  search = "search",
  speaker_notes = "speaker_notes",
}

export enum LogoNames {
  brooklyn = "brooklyn",
  logo_brooklyn = "logo_brooklyn",
  logo_nypl = "logo_nypl",
  logo_nypl_negative = "logo_nypl_negative",
  logo_queens = "logo_queens",
  lpa_negative = "lpa_negative",
  lpa_positive = "lpa_positive",
  nypl_circle_negative = "nypl_circle_negative",
  nypl_circle_positive = "nypl_circle_positive",
  nypl_full_lock_up_negative = "nypl_full_lock_up_negative",
  nypl_full_lock_up = "nypl_full_lock_up",
  nypl_no_lion_negative = "nypl_no_lion_negative",
  nypl_no_lion_positive = "nypl_no_lion_positive",
  queens_long = "queens_long",
  queens_stacked = "queens_stacked",
  schomburg_circle = "schomburg_circle",
  schomburg_positive = "schomburg_positive",
  simplye_black = "simplye_black",
  simplye_color = "simplye_color",
  snfl_negative = "snfl_negative",
  snfl_positive = "snfl_positive",
}

export enum IconColors {
  ui_black = "ui-black",
  ui_white = "ui-white",
  brand_primary = "brand-primary",
  brand_secondary = "brand-secondary",
  section_whats_on_primary = "section-whats-on-primary",
  section_books_and_more_primary = "section-books-and-more-primary",
  section_research_primary = "section-research-primary",
  section_research_secondary = "section-research-secondary",
  section_locations_primary = "section-locations-primary",
  section_research_library_lpa = "section-research-library-lpa",
  section_research_library_schomburg = "section-research-library-schomburg",
  section_research_library_schwartzman = "section-research-library-schwartzman",
}

export enum IconSizes {
  small = "small",
  medium = "medium",
  large = "large",
  xlarge = "xlarge",
  xxlarge = "xxlarge",
  xxxlarge = "xxxlarge",
}
