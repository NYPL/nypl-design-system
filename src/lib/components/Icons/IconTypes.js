"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.IconRotationTypes = void 0;
var IconRotationTypes;
(function (IconRotationTypes) {
    IconRotationTypes["rotate180"] = "rotate-180";
    IconRotationTypes["rotate270"] = "rotate-270";
    IconRotationTypes["rotate0"] = "rotate-0";
    IconRotationTypes["rotate90"] = "rotate-90";
})(
    (IconRotationTypes =
        exports.IconRotationTypes || (exports.IconRotationTypes = {}))
);
