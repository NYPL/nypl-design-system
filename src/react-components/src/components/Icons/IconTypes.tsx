export enum iconRotationOptions {
    rotate180 = "rotate-180",
    rotate270 = "rotate-270",
    rotate0 = "rotate-0",
    rotate90 = "rotate-90",
}
