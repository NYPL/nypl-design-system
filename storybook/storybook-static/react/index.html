<!doctype html><html lang="en"><head><meta charset="utf-8"><title>Storybook</title><meta name="viewport" content="width=device-width,initial-scale=1"><style>html, body {
    overflow: hidden;
    height: 100%;
    width: 100%;
    margin: 0;
    padding: 0;
  }</style><script>/* globals window */
  /* eslint-disable no-underscore-dangle */
  try {
    if (window.top !== window) {
      window.__REACT_DEVTOOLS_GLOBAL_HOOK__ = window.top.__REACT_DEVTOOLS_GLOBAL_HOOK__;
    }
  } catch (e) {
    // eslint-disable-next-line no-console
    console.warn('unable to connect to top frame for connecting dev tools');
  }</script><style>#root[hidden],
      #docs-root[hidden] {
        display: none !important;
      }</style></head><body><div id="root"></div><div id="docs-root"></div><script>window['DOCS_MODE'] = false;</script><script src="runtime~main.00bd9d033d7a0b972fc9.bundle.js"></script><script src="vendors~main.d2b6aafef1a259d165c0.bundle.js"></script><script src="main.3bf71672a6dd35499004.bundle.js"></script></body></html>