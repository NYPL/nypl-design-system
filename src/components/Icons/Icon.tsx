/* eslint-disable camelcase */
import * as React from "react";
import bem from "../../utils/bem";
import generateUUID from "../../helpers/generateUUID";

import {
  IconRotationTypes,
  IconNames,
  LogoNames,
  IconColors,
  IconSizes,
} from "./IconTypes";

import iconSvgs from "./IconSvgs";

export interface IconProps {
  /** Additional attributes to pass to the `<svg>` tag. */
  attributes?: { [key: string]: any };
  /** BlockName for use with BEM. See how to work with blockNames and BEM here: http://getbem.com/introduction/ */
  blockName?: string;
  /** Modifiers array for use with BEM. See how to work with modifiers and BEM here: http://getbem.com/introduction/ */
  modifiers?: string[];
  /** className that appears in addition to "icon" */
  className?: string;
  /** Icons designated as decorative will be ignored by screenreaders */
  decorative?: boolean;
  /** This text will be added as a child `<title>` element inside the `<svg>` tag.  It is recommended to do this for increased accessibility. */
  titleText?: string;
  /** Rotates the icon clockwise in increments of 90deg */
  iconRotation?: IconRotationTypes;
  /** Overrides default icon color (black). */
  color?: IconColors;
  /** Sets the icon size. */
  size?: IconSizes;
  /** The name of the icon you want to use. */
  name?: IconNames | LogoNames;
}

/**
 * Renders SVG-based icons.
 */
export default function Icon(props: React.PropsWithChildren<IconProps>) {
  const {
    blockName,
    decorative = false,
    className,
    titleText,
    iconRotation,
    color,
    size,
    modifiers = [],
    name,
    children,
    attributes = [],
  } = props;

  if (iconRotation) {
    modifiers.push(iconRotation);
  }

  if (color) {
    modifiers.push(color);
  }

  if (size) {
    modifiers.push(size);
  }

  const iconID = generateUUID();
  const iconProps = {
    className: bem("icon", modifiers, blockName, [className]),
    id: iconID,
    role: "img",
    title: titleText || null,
    "aria-hidden": decorative,
    ...attributes,
  };

  // Apply icon props to SVG that was passed as a child.
  const renderChildren = () =>
    React.Children.map(children, child =>
      React.cloneElement(child as JSX.Element, { ...iconProps })
    );

  // Component prop validation
  if (name && children) {
    throw new Error("Icon accepts either a name or children, not both");
  } else if (!name && !children) {
    console.warn(
      "Pass a name or any children to Icon to ensure an icon appears"
    );
  }

  // The user wants to render an existing icon. Load the icon and render it
  // as a component. Otherwise, we're just going to render the children that
  // were passed to this component.
  if (name) {
    const SvgComponent = iconSvgs[name];
    return <SvgComponent {...iconProps} />;
  }

  return <>{renderChildren()}</>;
}
