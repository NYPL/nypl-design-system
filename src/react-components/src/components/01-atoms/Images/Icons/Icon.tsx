
import * as React from "react";
import Icons from "@nypl/design-system-icons/dist/sprite/sprite.svg";
import bem from "../../../../utils/bem";

  // Wrapper Class for Icon
  export interface IconProps {
    name: string;
    blockName?: string;
    modifiers?: string[];
    decorative: boolean;
    role?: string;
    title?: boolean;
    desc?: boolean;
  }

  export default class Icon extends React.Component<IconProps, {}> {
    constructor(props: IconProps) {
      super(props);
    }

    render(): JSX.Element {
      let { modifiers, blockName, name, decorative, role, title, desc } = this.props;
      let icon_base_class = "icon";

      let iconProps = {
        "className": bem(icon_base_class, modifiers, blockName),
        "role": decorative ? "img" : role,
        "aria-hidden": decorative,
        "aria-labelledby": title ? "title-" + name : undefined,
        "aria-describedby": desc ? "desc-" + name : undefined
      };

      return (
        <svg {...iconProps} >
            {title &&
              <title id={`title-${name}`}>{ title }</title>
            }
            {desc &&
              <desc id={`title-${name}`}>{ desc }</desc>
            }
            <use href={`${Icons}#${name}`} />
        </svg>
      );
    }
  }
