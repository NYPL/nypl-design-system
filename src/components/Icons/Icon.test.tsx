import { expect } from "chai";
import * as Enzyme from "enzyme";
import * as React from "react";

import Icon from "./Icon";
import { IconNames } from "./IconTypes";

describe("Icon", () => {
  it("throws an error if both name and children are passed to Icon", () => {
    expect(() =>
      Enzyme.mount(
        <Icon name={IconNames.download} decorative={true}>
          <svg viewBox="0 0 24 14" xmlns="http://www.w3.org/2000/svg">
            <path
              fillRule="evenodd"
              clipRule="evenodd"
              d="M10.526 12.871L.263 1.676 1.737.324 12 11.52 22.263.324l1.474 1.352L13.474 12.87a2 2 0 01-2.948 0z"
            />
          </svg>
        </Icon>
      )
    ).to.throw("Icon accepts either a name or children, not both");
  });

  it("renders an icon based on the icon `name` prop", () => {
    const icon = Enzyme.mount(
      <Icon name={IconNames.download} decorative={true} />
    );
    expect(icon.children()).to.have.lengthOf(1);
  });

  it("renders an icon based on the child", () => {
    const icon = Enzyme.mount(
      <Icon decorative={true}>
        <svg viewBox="0 0 24 14" xmlns="http://www.w3.org/2000/svg">
          <path
            fillRule="evenodd"
            clipRule="evenodd"
            d="M10.526 12.871L.263 1.676 1.737.324 12 11.52 22.263.324l1.474 1.352L13.474 12.87a2 2 0 01-2.948 0z"
          />
        </svg>
      </Icon>
    );
    expect(icon.children()).to.have.lengthOf(1);
  });
});
