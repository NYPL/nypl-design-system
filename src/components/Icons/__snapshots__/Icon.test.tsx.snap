// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Icon renders the UI snapshot correctly 1`] = `
<svg
  aria-hidden={true}
  className="chakra-icon css-1grhd2q"
  focusable={false}
  id="basic"
  role="img"
  title="download icon"
  viewBox="0 0 24 24"
>
  <g
    stroke="currentColor"
    strokeWidth="1.5"
  >
    <path
      d="M9,9a3,3,0,1,1,4,2.829,1.5,1.5,0,0,0-1,1.415V14.25"
      fill="none"
      strokeLinecap="round"
    />
    <path
      d="M12,17.25a.375.375,0,1,0,.375.375A.375.375,0,0,0,12,17.25h0"
      fill="currentColor"
      strokeLinecap="round"
    />
    <circle
      cx="12"
      cy="12"
      fill="none"
      r="11.25"
      strokeMiterlimit="10"
    />
  </g>
</svg>
`;

exports[`Icon renders the UI snapshot correctly 2`] = `
<div
  className="css-0"
>
  <svg
    aria-hidden={true}
    className="css-0"
    id="customIcon"
    role="img"
    title="undefined icon"
    viewBox="0 0 24 14"
    xmlns="http://www.w3.org/2000/svg"
  >
    <path
      clipRule="evenodd"
      d="M10.526 12.871L.263 1.676 1.737.324 12 11.52 22.263.324l1.474 1.352L13.474 12.87a2 2 0 01-2.948 0z"
      fillRule="evenodd"
    />
  </svg>
</div>
`;

exports[`Icon renders the UI snapshot correctly 3`] = `
<svg
  aria-hidden={true}
  className="chakra-icon css-7rufou"
  focusable={false}
  id="chakra"
  role="img"
  title="download icon"
  viewBox="0 0 24 24"
>
  <g
    stroke="currentColor"
    strokeWidth="1.5"
  >
    <path
      d="M9,9a3,3,0,1,1,4,2.829,1.5,1.5,0,0,0-1,1.415V14.25"
      fill="none"
      strokeLinecap="round"
    />
    <path
      d="M12,17.25a.375.375,0,1,0,.375.375A.375.375,0,0,0,12,17.25h0"
      fill="currentColor"
      strokeLinecap="round"
    />
    <circle
      cx="12"
      cy="12"
      fill="none"
      r="11.25"
      strokeMiterlimit="10"
    />
  </g>
</svg>
`;

exports[`Icon renders the UI snapshot correctly 4`] = `
<svg
  aria-hidden={true}
  className="chakra-icon css-1grhd2q"
  data-testid="props"
  focusable={false}
  id="props"
  role="img"
  title="download icon"
  viewBox="0 0 24 24"
>
  <g
    stroke="currentColor"
    strokeWidth="1.5"
  >
    <path
      d="M9,9a3,3,0,1,1,4,2.829,1.5,1.5,0,0,0-1,1.415V14.25"
      fill="none"
      strokeLinecap="round"
    />
    <path
      d="M12,17.25a.375.375,0,1,0,.375.375A.375.375,0,0,0,12,17.25h0"
      fill="currentColor"
      strokeLinecap="round"
    />
    <circle
      cx="12"
      cy="12"
      fill="none"
      r="11.25"
      strokeMiterlimit="10"
    />
  </g>
</svg>
`;
